File: src/core/runtime.zig
const std = @import("std");

pub const Runtime = struct {
    allocator: std.mem.Allocator,
    io: std.Io,
    environ_map: *std.process.Environ.Map,
    fs_cwd: ?[]const u8 = null,
    stdout_sink: ?OutputSink = null,
    stderr_sink: ?OutputSink = null,
};

pub const OutputTarget = enum {
    stdout,
    stderr,
};

pub const OutputSink = struct {
    context: *anyopaque,
    writeFn: *const fn (context: *anyopaque, bytes: []const u8) void,

    pub fn write(self: @This(), bytes: []const u8) void {
        self.writeFn(self.context, bytes);
    }
};

var current_runtime: ?Runtime = null;

pub fn isInitialized() bool {
    return current_runtime != null;
}

pub fn set(runtime: Runtime) void {
    current_runtime = runtime;
}

fn getMutable() *Runtime {
    if (current_runtime == null) {
        std.debug.panic("mg runtime is not initialized", .{});
    }
    return &current_runtime.?;
}

pub fn get() *const Runtime {
    if (current_runtime == null) {
        std.debug.panic("mg runtime is not initialized", .{});
    }
    return &current_runtime.?;
}

pub fn getFsCwd() ?[]const u8 {
    return get().fs_cwd;
}

pub fn swapFsCwd(path: ?[]const u8) ?[]const u8 {
    const runtime = getMutable();
    const previous = runtime.fs_cwd;
    runtime.fs_cwd = path;
    return previous;
}

pub fn swapOutputSink(target: OutputTarget, sink: ?OutputSink) ?OutputSink {
    const runtime = getMutable();
    const slot = switch (target) {
        .stdout => &runtime.stdout_sink,
        .stderr => &runtime.stderr_sink,
    };
    const previous = slot.*;
    slot.* = sink;
    return previous;
}

fn writeToFile(file: std.Io.File, bytes: []const u8) void {
    var writer_buf: [256]u8 = undefined;
    var writer = file.writer(get().io, &writer_buf);
    writer.interface.writeAll(bytes) catch {};
    writer.interface.flush() catch {};
}

fn writeOutput(target: OutputTarget, bytes: []const u8) void {
    if (bytes.len == 0) return;

    const sink = switch (target) {
        .stdout => get().stdout_sink,
        .stderr => get().stderr_sink,
    };
    if (sink) |captured| {
        captured.write(bytes);
        return;
    }

    const file = switch (target) {
        .stdout => std.Io.File.stdout(),
        .stderr => std.Io.File.stderr(),
    };
    writeToFile(file, bytes);
}

pub fn writeStdout(bytes: []const u8) void {
    writeOutput(.stdout, bytes);
}

pub fn writeStderr(bytes: []const u8) void {
    writeOutput(.stderr, bytes);
}

File: src/core/runtime.cpp
#include <iostream>
#include <stdexcept>

#include <mg/core/runtime.hpp>

namespace mg
{
namespace
{
auto global_runtime = std::optional<Runtime> {};

[[nodiscard]] std::ostream& stream_for(OutputTarget target)
{
  auto& runtime = current_runtime();
  if (target == OutputTarget::stdout_stream) {
    return runtime.out == nullptr ? std::cout : *runtime.out;
  }

  return runtime.err == nullptr ? std::cerr : *runtime.err;
}

void write_output(OutputTarget target, std::string_view bytes)
{
  if (bytes.empty()) {
    return;
  }

  stream_for(target) << bytes;
  stream_for(target).flush();
}
}  // namespace

void set_runtime(Runtime runtime)
{
  global_runtime = std::move(runtime);
}

bool runtime_initialized() noexcept
{
  return global_runtime.has_value();
}

Runtime& current_runtime()
{
  if (!global_runtime) {
    global_runtime = Runtime {
        .out = &std::cout,
        .err = &std::cerr,
    };
  }

  return *global_runtime;
}

std::optional<std::filesystem::path> get_fs_cwd()
{
  return current_runtime().fs_cwd;
}

std::optional<std::filesystem::path> swap_fs_cwd(
    std::optional<std::filesystem::path> path)
{
  auto& runtime = current_runtime();
  auto previous = runtime.fs_cwd;
  runtime.fs_cwd = std::move(path);
  return previous;
}

void write_stdout(std::string_view bytes)
{
  write_output(OutputTarget::stdout_stream, bytes);
}

void write_stderr(std::string_view bytes)
{
  write_output(OutputTarget::stderr_stream, bytes);
}
}  // namespace mg
